### PR TITLE
feature/maximum iteration per rank to prevent infinite loop

### PR DESCRIPTION
**Pull request by Jerry-Jinfeng-Guo**
- [x] maximum tap ranges per rank
- [x] 2x max tap range as max iter for rank
- [x] unit test
- [x] new error

### CODE REVIEW

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp [hunk 1]
```diff
@@ -665,8 +672,13 @@ class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator,
                 if (tap_changed) {
                     break;
                 }
+                iterations_per_rank[++rank_index] = 0;
             }
             if (tap_changed) {
+                if (++iterations_per_rank[rank_index] > 2 * regulator_order[rank_index][0].transformer.tap_range()) {
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:678]: should be (pseudocode):
```
max_iters[rank_k] = max( tap_range() for transformer in regulator_order[rank_idx] )
```
because trafo 1 in the same rank may have a larger tap range than trafo 0.

probably a good idea to pre-calculate (maybe even in a separate function)

File: power_grid_model_c/power_grid_model/include/power_grid_model/common/exception.hpp [hunk 2]
```diff
@@ -112,6 +112,11 @@ class IterationDiverge : public PowerGridError {
     }
 };
 
+class MaxIterationReached : public PowerGridError {
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/common/exception.hpp:115]: shouldn't this derive from `IterationDiverge`?

```suggestion
class MaxIterationReached : public IterationDiverge {
```

Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/common/exception.hpp:115]: please also add to the python exceptions incl parsing

Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/common/exception.hpp:115]: do we also want to update the `MathSolvers`?

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp [hunk 3]
```diff
@@ -665,8 +672,13 @@ class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator,
...
+                    throw MaxIterationReached{
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:679]: i believe that if the optimizer throws, that the cache is not restored. this is a bug in the current implementation because it will already run into this edge case when a `SparseMatrixError` or `IterationDiverge` happens in the solver and it is not able to recover by re-running with the fallback method. Will create a Jira ticket

File: power_grid_model_c/power_grid_model/include/power_grid_model/common/exception.hpp (hunk 2 above)
Comment (Jerry-Jinfeng-Guo) [reply power_grid_model_c/power_grid_model/include/power_grid_model/common/exception.hpp:115]: After discussion: original `IterationDiverge` is probably not terribly accurate, but for the sake of keeping minimum change to the user side, `MaxIteration` error should derive from it and only mention the maximum iteration reached situation. This has nothing to do with convergence per se.

Comment (Jerry-Jinfeng-Guo) [reply power_grid_model_c/power_grid_model/include/power_grid_model/common/exception.hpp:115]: See discussion above.

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp (hunk 1 above)
Comment (Jerry-Jinfeng-Guo) [reply power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:678]: Conclusion: maximum tap range should be taken

File: power_grid_model_c/power_grid_model/include/power_grid_model/common/exception.hpp [hunk 4]
```diff
@@ -106,12 +106,18 @@ class NotObservableError : public PowerGridError {
 
 class IterationDiverge : public PowerGridError {
   public:
+    IterationDiverge() {}
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/common/exception.hpp:109]:
```suggestion
    IterationDiverge() = default;
```

File: power_grid_model_c/power_grid_model/include/power_grid_model/common/exception.hpp (hunk 2 above)
Comment (mgovers) [reply power_grid_model_c/power_grid_model/include/power_grid_model/common/exception.hpp]: did you forget to push this by mistake?

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp [hunk 5]
```diff
@@ -354,6 +354,9 @@ template <transformer_c... TransformerTypes> class TransformerWrapper {
     IntS tap_max() const {
         return apply([](auto const& t) { return t.tap_max(); });
     }
+    IntS tap_range() const {
+        return apply([](auto const& t) { return std::abs(t.tap_max() - t.tap_min()); });
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:358]: this will wrap around if `|tap_min - tap_max| >= 128`... not sure what would be a good solution though

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp [hunk 6]
```diff
@@ -665,8 +673,14 @@ class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator,
                 if (tap_changed) {
                     break;
                 }
+                iterations_per_rank[++rank_index] = 0;
             }
             if (tap_changed) {
+                if (static_cast<unsigned long>(++iterations_per_rank[rank_index]) >
+                    2 * max_tap_ranges_per_rank[rank_index]) {
+                    throw MaxIterationReached{
+                        "TapPositionOptimizer::iterate, maximum iterations reached, no solution."};
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:682]: it may be useful to add the `maximum iterations reached` to the constructor of the exception. that will also help when making the python exception parser

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp [hunk 7]
```diff
@@ -736,6 +750,9 @@ class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator,
             // min voltage at control side => max voltage at tap side => max tap pos
             return transformer.tap_max();
         };
+        constexpr auto tap_pos_range_cmp = [](RegulatedTransformer const& a, RegulatedTransformer const& b) {
+            return a.transformer.tap_range() < b.transformer.tap_range();
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:754]: probably this is a good location to take care of the wraparound

File: tests/cpp_unit_tests/test_tap_position_optimizer.cpp [hunk 8]
```diff
@@ -997,6 +987,85 @@ TEST_CASE("Test Tap position optimizer") {
...
+        state_b.u_pu = [&state_b, &regulator_b](ControlSide side) {
+            CHECK(side == regulator_b.control_side());
+
+            // tap pos closer to tap_max at tap side <=> lower voltage at control side
+            return static_cast<DoubleComplex>(test::normalized_lerp(state_b.tap_pos, state_b.tap_max, state_b.tap_min));
+        };
+
+        auto update_data = TransformerTapRegulatorUpdate{.id = 4, .u_set = 0.5, .u_band = 0.0};
+
+        // TODO(jguo): final sub test fail -> move to stand alone test to catch error
```
Comment (mgovers) [inline tests/cpp_unit_tests/test_tap_position_optimizer.cpp:1042]: what do you mean?

File: tests/cpp_unit_tests/test_tap_position_optimizer.cpp [hunk 9]
```diff
@@ -997,6 +987,85 @@ TEST_CASE("Test Tap position optimizer") {
...
+        SUBCASE("no valid value in band") {
+            state_b.tap_min = 1;
+            state_b.tap_max = 5;
+            state_b.tap_pos = 3;
```
Comment (mgovers) [inline tests/cpp_unit_tests/test_tap_position_optimizer.cpp:1046]: due to the wraparound case: may be good to add an edge case with really large `tap_max - tap_min`

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp (hunk 5 above)
Comment (Jerry-Jinfeng-Guo) [reply power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:358]: could use some upgrade ;)

File: tests/cpp_unit_tests/test_tap_position_optimizer.cpp (hunk 9 above)
Comment (Jerry-Jinfeng-Guo) [reply tests/cpp_unit_tests/test_tap_position_optimizer.cpp:1046]: the upgrade renders this test unnecessary.

File: tests/cpp_unit_tests/test_tap_position_optimizer.cpp (hunk 8 above)
Comment (Jerry-Jinfeng-Guo) [reply tests/cpp_unit_tests/test_tap_position_optimizer.cpp:1042]: meaning this was originally left for today to patch up but I found this to be minimum work yesterday and did it after all (forgetting a couple things of course...)

File: tests/cpp_unit_tests/test_tap_position_optimizer.cpp [hunk 10]
```diff
@@ -1004,6 +994,83 @@ TEST_CASE("Test Tap position optimizer") {
...
+
+        auto update_data = TransformerTapRegulatorUpdate{.id = 4, .u_set = 0.5, .u_band = 0.0};
+
+        SUBCASE("no valid value in band") {
+            state_b.tap_min = 1;
+            state_b.tap_max = 5;
+            state_b.tap_pos = 3;
+
+            update_data.u_set = 0.4;
+            update_data.u_set = 0.01;
```
Comment (nitbharambe) [inline tests/cpp_unit_tests/test_tap_position_optimizer.cpp:1054]: Seems like you missed this

Comment (Jerry-Jinfeng-Guo) [reply tests/cpp_unit_tests/test_tap_position_optimizer.cpp:1054]: Oops. Either is valid

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp [hunk 11]
```diff
@@ -578,6 +583,7 @@ class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator,
     using TransformerRanker = TransformerRanker_;
 
   private:
+    mutable std::vector<uint64_t> max_tap_ranges_per_rank{};
```
Comment (nitbharambe) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:586]: tap_min and tap_max are not going to change for a input_data.
I wonder if we can just have tap_range at `RegulatedObjects` instead. They can be calculated in `retrieve_regulator_info()`.
And then the max iteration of each rank in `RankedTransformerGroups`. They can be calculated at `rank_transformers()`

Comment (Jerry-Jinfeng-Guo) [reply power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:586]: > tap_min and tap_max are not going to change for a input_data.
I wonder if we can just have tap_range at RegulatedObjects instead. They can be calculated in retrieve_regulator_info().

`tap_range` is already available in transformers as a function call. In terms of number of calls, they are the same whether they live as a member or as a function call, because only once will they be required throughout the code (here). If we put everything inside `RegulatedObjects`, which is container of indices, I am not sure if this is making things less scattered or more so XD

> And then the max iteration of each rank in RankedTransformerGroups. They can be calculated at rank_transformers()

Having this per rank information inside `RankedTransformerGroups` indeed makes sense, and that was my original code as well. After bumping into some template mismatch, I figure it might also make sense that this is a per optimizer requirement, and not necessarily order group property (which is again groups of indices).

File: tests/cpp_unit_tests/test_tap_position_optimizer.cpp [hunk 12]
```diff
@@ -1004,6 +994,81 @@ TEST_CASE("Test Tap position optimizer") {
...
+            state, 4,
+            TransformerTapRegulatorInput{.id = 4,
+                                         .regulated_object = 2,
+                                         .status = 1,
+                                         .control_side = ControlSide::side_2,
+                                         .u_set = 0.0,
+                                         .u_band = 0.0,
+                                         .line_drop_compensation_r = 0.0,
+                                         .line_drop_compensation_x = 0.0},
+            transformer_b.math_model_type(), 1.0);
```
Comment (nitbharambe) [inline tests/cpp_unit_tests/test_tap_position_optimizer.cpp:1028]: Would be nice to avoid repetition here

**Review by nitbharambe — APPROVED** [review]

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp [hunk 13]
```diff
@@ -666,8 +676,13 @@ class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator,
                 if (tap_changed) {
                     break;
                 }
+                iterations_per_rank[++rank_index] = 0; // NOSONAR
```
Comment (nitbharambe) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:679]: out of curiosity, what does sonar say?

Comment (Jerry-Jinfeng-Guo) [reply power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:679]: something like "omg, more than one thing at a line"

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp (hunk 11 above)
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:586]: i know it's already been merged and it doesn't necessarily need active changing, but what's the reasons you need `mutable`?

using `mutable` is usually dangerous because it overrides **any** `const` statement. that means that a statement like

```cpp
#include <iostream>

class A{
 private:
  int regular_a{};
  mutable int mutable_a{};

 public:
  int get_a() const {
    // return a++;  // compile error: instance is const
    return regular_a;       // OK
  }

  int get_mutable_a() const {
    return mutable_a++;  // OK: mutable overrides const
  }
};

void print_a(const A& const_a) {
    std::cout << "a: ";
    for (auto _ : {0, 1, 2}) {
        std::cout << const_a.get_a() << ", ";
    };
    std::cout << "\n";
}
void print_mutable_a(const A& const_a) {
    std::cout << "mutable a: ";
    for (auto _ : {0, 1, 2}) {
        std::cout << const_a.get_mutable_a() << ", ";
    };
    std::cout << "\n";
}

int main() {
    A const const_a{};
    print_a(const_a);
    print_mutable_a(const_a);
    return 0;
}
```
with output:
```txt
a: 0, 0, 0
mutable_a: 0, 1, 2
```

using `mutable` is usually only useful when dealing with things like loggers or other boilerplate code with intentional side effects on otherwise constant objects. in most other cases, changing the `const`ness in the function declaration or changing the order of function calls is usually a better way

**Review by mgovers — COMMENTED** [review]
just want to understand one specific part of the code change. not overruling the approval in any way cause i do like the general solution

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp (hunk 11 above)
Comment (Jerry-Jinfeng-Guo) [reply power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:586]: The `mutable` was used due to the fact that the only place that it seemingly fit is in the `initialize` function. Which is not actually the case because the function name is misleading. Since for now it's a private member and only exposed to this misleading function, active changing is indeed not needed. In the possible following-up refactoring, we could work a bit on the `initialize` and `exploit_neighbourhood` and move the max iteration mechanism somewhere proper.

Comment (Jerry-Jinfeng-Guo) [reply power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:586]: Hahaha, that's the option I consider even more frightening XD

Comment (Jerry-Jinfeng-Guo) [reply power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:586]: See https://github.com/PowerGridModel/power-grid-model/pull/617
